class/class.midtrans-gateway-subscription.php

```diff
@@ -0,0 +1,225 @@
+      } catch (Exception $e) {
+        $this->setLogError( $e->getMessage() );
+        WC_Midtrans_Utils::json_print_exception( $e, $this );
+        exit();
+      }
+
+      if(property_exists($this,'enable_redirect') && $this->enable_redirect == 'yes'){
+        $redirectUrl = $snapResponse->redirect_url;
+      }else{
+        $redirectUrl = $order->get_checkout_payment_url( true )."&snap_token=".$snapResponse->token;
+      }
+
+      // Add snap token & snap redirect url to $order metadata.
+      $order->update_meta_data('_mt_payment_snap_token',$snapResponse->token);
+      $order->update_meta_data('_mt_payment_url',$snapResponse->redirect_url);
+      $order->save();
+      
+      if(property_exists($this,'enable_immediate_reduce_stock') && $this->enable_immediate_reduce_stock == 'yes'){
+        wc_reduce_stock_levels($order);
+      }
+
+      $successResponse['redirect'] = $redirectUrl;
+      return $successResponse;
+    }
+
+    /**
+     * Hook function that will be called on receipt page.
+     * Output HTML for Snap payment page. Including `snap.pay()` part.
+     * @param  string $order_id generated by WC.
+     * @return string HTML.
+     */
+    function receipt_page( $order_id ) {
+      global $woocommerce;
+      $pluginName = 'subscription';
+      require_once(dirname(__FILE__) . '/payment-page.php'); 
+    }
+
+    /**
+     * Scheduled_subscription_payment function.
+     * This function is called when renewal order is triggered.
+     * 
+     * @param float $amount_to_charge float The amount to charge.
+     * @param WC_Order $renewal_order A WC_Order object created to record the renewal payment.
+     */
+    public function scheduled_subscription_payment( $amount_to_charge, $renewal_order ) {
+      if (!$renewal_order) {  
+        return array ( 'result' => 'failure' );
+      }
+      $order_id = $renewal_order->get_id();
+      $subscription_order = wcs_get_subscriptions_for_renewal_order( $order_id );
+      $checkout_payment_url = $renewal_order->get_checkout_payment_url();
+
+      // Retrieve card token from meta.
+      foreach ( $subscription_order as $subscription ) {
+        $card_token = $subscription->get_meta('_mt_subscription_card_token');
+      }
+
+      // if card_token null, the transaction will fail.
+      if ($card_token == '' ) {
+        $renewal_order->update_status( 'failed', __('Midtrans subscription payment failed.', 'midtrans-woocommerce') );
+        $renewal_order->add_order_note( __( 'Customer didn\'t tick the SAVE CARD on previous payment. <br> Please click <a href="'.$checkout_payment_url.'">here</a> to renew the payment.', 'midtrans-woocommerce' ), 1 );
+        return array('result' => 'failure');
+        exit();
+      }
+
+      // Get data for charging to Midtrans API.
+      $params = $this->getPaymentRequestData( $order_id );
+      $params['credit_card']['secure'] = false;
+      $params['credit_card']['token_id'] = $card_token;
+
```

Comment:
On Snap payload, is it possible to please add `"custom_field3": "woocommerce-subscription-renewal"` so it can be easily identified on Midtrans Dashboard.

Reply:
done on https://github.com/veritrans/SNAP-Woocommerce/pull/39/commits/ae189bdb9e8860c8a59516c82c338c4fea18a4e6